src/Common/QuillLogger.h

```diff
@@ -0,0 +1,160 @@
+#pragma once
+
+#include <Common/QuillLogger_fwd.h>
+#include <Common/ThreadPool_fwd.h>
+
+#include <quill/core/Common.h>
+#include <quill/Logger.h>
+#include <quill/core/LogLevel.h>
+#include <quill/sinks/FileSink.h>
+#include <quill/sinks/StreamSink.h>
+#include <quill/Frontend.h>
+
+#include <deque>
+#include <filesystem>
+#include <future>
+
+using FreeThreadPool = ThreadPoolImpl<std::thread>;
+
+namespace DB
+{
+
+struct QuillFrontendOptions
+{
+    static constexpr quill::QueueType queue_type = quill::QueueType::UnboundedBlocking;
+
+    /**
+   * Initial capacity of the queue. Used for UnboundedBlocking, UnboundedDropping, and
+   * UnboundedUnlimited. Also serves as the capacity for BoundedBlocking and BoundedDropping.
+   */
+    static constexpr uint32_t initial_queue_capacity = 128 * 1024;
+
+    /**
+   * Interval for retrying when using BoundedBlocking or UnboundedBlocking.
+   * Applicable only when using BoundedBlocking or UnboundedBlocking.
+   */
+    static constexpr uint32_t blocking_queue_retry_interval_ns = 800;
+
+    /**
+   * Enables huge pages on the frontend queues to reduce TLB misses. Available only for Linux.
+   */
+    static constexpr quill::HugePagesPolicy huge_pages_policy = quill::HugePagesPolicy::Never;
+};
+
+using QuillLoggerPtr = quill::LoggerImpl<QuillFrontendOptions> *;
+
+void startQuillBackend();
+
+quill::LogLevel parseQuillLogLevel(std::string_view level);
+
+class ConsoleSink : public quill::StreamSink
+{
+public:
+    enum class Stream
+    {
+        STDOUT,
+        STDERR,
+    };
+
+    constexpr std::string_view streamToString(Stream stream)
+    {
+        switch (stream)
+        {
+            case Stream::STDOUT:
+                return "stdout";
+            case Stream::STDERR:
+                return "stderr";
+        }
+    }
+
+    explicit ConsoleSink(Stream stream, bool enable_colors_ = false);
+
+    ~ConsoleSink() override = default;
+
+    QUILL_ATTRIBUTE_HOT void write_log(
```

Comment:
Does `QUILL_ATTRIBUTE_HOT` do anything for us? Seems it's GCC only. Although clang supports it it's explicitly disabled in the definition (`contrib/quill/include/quill/core/Attributes.h`)